src/indexed_crate.rs

```diff
@@ -41,6 +41,166 @@ pub struct IndexedCrate<'a> {
     pub(crate) manually_inlined_builtin_traits: HashMap<Id, Item>,
 }
 
+/// Map a Key to a List (Vec) of values
+///
+/// It also has some nice operations for pushing a value to the list, or extending the list with
+/// many values.
+struct MapList<K, V>(HashMap<K, Vec<V>>);
+
+impl<K: std::cmp::Eq + std::hash::Hash, V> FromIterator<(K, V)> for MapList<K, V> {
+    #[inline]
+    fn from_iter<T: IntoIterator<Item = (K, V)>>(iter: T) -> Self {
+        let iter = iter.into_iter();
+        let (min, max) = iter.size_hint();
+        // Reserving this space doesn't give a measurable gain, but neither does it hurt so we are
+        // keeping it around
+        let mut map = Self::with_capacity(max.unwrap_or(min));
+        for (key, value) in iter {
+            map.insert(key, value);
+        }
+        map
+    }
+}
+
+impl<K: std::cmp::Eq + std::hash::Hash, V> Extend<(K, V)> for MapList<K, V> {
+    #[inline]
+    fn extend<T: IntoIterator<Item = (K, V)>>(&mut self, iter: T) {
+        // We could use Iterator::size_hint here to reserve some space, but I measured a 2%-3%
+        // regression when doing that.
+        for (key, value) in iter.into_iter() {
+            self.insert(key, value);
+        }
+    }
+}
+
+impl<K: std::cmp::Eq + std::hash::Hash, V> MapList<K, V> {
+    #[inline]
+    pub fn with_capacity(capacity: usize) -> Self {
+        Self(HashMap::with_capacity(capacity))
+    }
+
+    #[inline]
+    pub fn into_inner(self) -> HashMap<K, Vec<V>> {
+        self.0
+    }
+
+    #[inline]
+    pub fn insert(&mut self, key: K, value: V) {
+        match self.0.entry(key) {
+            Entry::Occupied(mut entry) => entry.get_mut().push(value),
+            Entry::Vacant(entry) => {
+                entry.insert(vec![value]);
+            }
+        }
+    }
+}
+
+/// Build the imports index
+///
+/// When compiled using the `rayon` feature, build it in parallel. Specifically, this paralelizes
+/// getting the [`IndexedCrate::publicly_importable_names`] over the items in the index.
+fn build_imports_index<'a>(
+    index: &'a HashMap<Id, Item>,
+    value: &IndexedCrate<'a>,
```

Comment:
This `&IndexedCrate<'a>` is actually not fully constructed at the time of this call. That is a potential footgun that could waste a few hours in a future iteration of tweaking this code, and I'd prefer to avoid introducing it in the first place.

Would you mind inlining this method back into `IndexedCrate::new()` where it's much more clear that the `value` is not fully constructed yet?

It might even be good to add an explicit comment about how it's not fully constructed at that point of use, to explain why extracting that complexity into its own method isn't desirable. In retrospect, that isn't obvious and I totally understand why you moved it here as a result 😅

Reply:
Fixed!